trieste/utils/pareto.py

```diff
@@ -101,10 +110,161 @@ def _bounds_2d(front: TensorType) -> BoundedVolumes:
+        )
+
+        return BoundedVolumes(lower_result, upper_result)
+
+    @staticmethod
+    def _is_test_required(smaller: TensorType) -> TensorType:
+        idx_dom_augm = tf.reduce_any(smaller, axis=1)
+        is_dom_augm = tf.reduce_all(idx_dom_augm)
+
+        return is_dom_augm
+
+    def _divide_conquer_nd(
+        self, front: TensorType, threshold: TensorType | float = 0
+    ) -> TensorType:
+        # Divide and conquer strategy to compute the cells covering the non-dominated region.
+        # Generic version: works for an arbitrary number of objectives.
+
+        len_front, number_of_objectives = front.shape
+        jitter = gpflow.config.default_jitter()
+        lower_result = tf.zeros([0, number_of_objectives], dtype=tf.int32)
+        upper_result = tf.zeros([0, number_of_objectives], dtype=tf.int32)
 
-        return BoundedVolumes(lower, upper)
+        min_front = tf.reduce_min(front, axis=0, keepdims=True) - 1
+        max_front = tf.reduce_max(front, axis=0, keepdims=True) + 1
+        pseudo_front = tf.concat([min_front, front, max_front], axis=0)
+
+        pseudo_front_idx = tf.concat(
+            [
+                tf.zeros([1, number_of_objectives], dtype=tf.int32),
+                tf.argsort(front, axis=0) + 1,  # +1 as index zero is reserved for the ideal point
+                tf.ones([1, number_of_objectives], dtype=tf.int32) * len_front + 1,
+            ],
+            axis=0,
+        )
+
+        divide_conquer_cells = tf.stack(
+            [
+                tf.zeros(number_of_objectives, dtype=tf.int32),
+                (int(pseudo_front_idx.shape[0]) - 1)
+                * tf.ones(number_of_objectives, dtype=tf.int32),
+            ],
+            axis=0,
+        )[None]
+
+        total_size = tf.reduce_prod(max_front - min_front)
+
+        def while_body(
+            divide_conquer_cells: TensorType,
+            lower_result: TensorType,
+            upper_result: TensorType,
+        ) -> Tuple[TensorType, TensorType, TensorType]:
+            divide_conquer_cells = tf.unstack(divide_conquer_cells, axis=0)
+            cell = divide_conquer_cells[-1]
+            divide_conquer_cells = tf.cond(
+                tf.not_equal(tf.size(divide_conquer_cells[:-1]), 0),
+                lambda: tf.stack(divide_conquer_cells[:-1]),
+                lambda: tf.zeros([0, 2, number_of_objectives], dtype=tf.int32),
+            )
+
+            arr = tf.range(number_of_objectives)
+            lower_idx = tf.gather_nd(pseudo_front_idx, tf.stack((cell[0], arr), -1))
+            upper_idx = tf.gather_nd(pseudo_front_idx, tf.stack((cell[1], arr), -1))
+            lower = tf.gather_nd(pseudo_front, tf.stack((lower_idx, arr), -1))
+            upper = tf.gather_nd(pseudo_front, tf.stack((upper_idx, arr), -1))
+
+            test_accepted = self._is_test_required((upper - jitter) < front)
+            lower_result, upper_result = tf.cond(
```

Comment:
here's another re-assignment

Reply:
you can also just add an underscore to variable names if they've already been used `lower_result_`. I wouldn't recommend it normally but I don't have time to suggest anything better

Reply:
I am thinking to use `lower_result_final` since it "final" wrt to the method which it resides. what do you think?

Reply:
sounds good - you understand the algorithm so are better placed to name things